File: testcases/benchmark/sorting/quick_sort.f90
recursive subroutine quick_sort(arr, low, high)
    integer, intent(inout) :: arr(:)
    integer, intent(in) :: low, high
    integer :: pivot, temp, i, j, pi

    if (low < high) then
        pivot = arr(high)
        i = low - 1
        do j = low, high - 1
            if (arr(j) <= pivot) then
                i = i + 1
                temp = arr(i)
                arr(i) = arr(j)
                arr(j) = temp
            end if
        end do
        temp = arr(i + 1)
        arr(i + 1) = arr(high)
        arr(high) = temp
        pi = i + 1
        call quick_sort(arr, low, pi - 1)
        call quick_sort(arr, pi + 1, high)
    end if
end subroutine quick_sort

program main
end program main

File: testcases/benchmark/sorting/quick_sort.rs
fn partition(arr: &mut Vec<i32>, low: usize, high: usize) -> usize {
    let pivot = arr[high];
    let mut i = low;
    for j in low..high {
        if arr[j] <= pivot {
            arr.swap(i, j);
            i += 1;
        }
    }
    arr.swap(i, high);
    i
}

fn quick_sort(arr: &mut Vec<i32>, low: usize, high: usize) {
    if low < high {
        let pi = partition(arr, low, high);
        if pi > 0 {
            quick_sort(arr, low, pi - 1);
        }
        quick_sort(arr, pi + 1, high);
    }
}

fn main() {}
